File: src/lib/kafe/src/main/java/kafe/BaseThen.java
package kafe;

import java.util.function.Function;

/**
 * BaseThen extends BaseCheck to support BDD pattern.
 * It reuses all Check functionality but with BDD naming.
 */
public abstract class BaseThen<R> extends BaseCheck<R> {
    public BaseThen(String name, Function<R, Object> thenCb) {
        super(name, thenCb);
    }
    
    // Alias verifyCheck to butThen for BDD pattern
    @Override
    public abstract Object verifyCheck(
        R store,
        Function<R, Object> checkCB,
        ITTestResourceConfiguration testResourceConfiguration,
        Object artifactory
    );
    
    // Alias test to butThenTest for BDD pattern
    public Object butThenTest(R store, ITTestResourceConfiguration testResourceConfiguration, String filepath) throws Exception {
        return super.test(store, testResourceConfiguration, filepath);
    }
    
    // Overload with artifactory
    public Object butThenTest(R store, ITTestResourceConfiguration testResourceConfiguration, String filepath, Object artifactory) throws Exception {
        return super.test(store, testResourceConfiguration, filepath, artifactory);
    }
}

File: src/lib/rubeno/lib/base_then.rb
module Rubeno
  class BaseThen < BaseCheck
    attr_accessor :adapter
    
    def initialize(name, then_cb)
      super(name, then_cb)
      @adapter = nil
    end
    
    def verify_check(store, check_cb, test_resource_configuration, artifactory = nil)
      # This should call the adapter's verify method if available
      if @adapter && @adapter.respond_to?(:verify)
        @adapter.verify(store, check_cb, test_resource_configuration, artifactory)
      else
        # Fallback to calling the check_cb directly
        if check_cb.respond_to?(:call)
          check_cb.call(store)
        else
          store
        end
      end
    end
    
    # Alias for BDD pattern - kept for compatibility but uses universal method internally
    def but_then(store, then_cb, test_resource_configuration, artifactory = nil)
      verify_check(store, then_cb, test_resource_configuration, artifactory)
    end
    
    # Override test to not require filepath for backward compatibility
    def test(store, test_resource_configuration, filepath = nil, artifactory = nil)
      super(store, test_resource_configuration, filepath, artifactory)
    end
  end
end
